sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/NettyAsyncHttpResponse.java

```diff
@@ -38,14 +38,15 @@ public NettyAsyncHttpResponse(HttpClientResponse reactorNettyResponse, Connectio
 
     @Override
     public Flux<ByteBuffer> getBody() {
-        return bodyIntern()
-            .map(byteBuf -> this.disableBufferCopy ? byteBuf.nioBuffer() : deepCopyBuffer(byteBuf))
-            .doFinally(ignored -> close());
+        return Flux.using(() -> this, response -> response.bodyIntern()
+            .map(byteBuf -> this.disableBufferCopy ? byteBuf.nioBuffer() : deepCopyBuffer(byteBuf)),
+            NettyAsyncHttpResponse::close);
```

Comment:
I think assumption here when deep-copy-disabled:true is - 
1. the Flux elements as processed in the emitting thread, and 
3. There won’t be any operator in the chain later with (implicit or explicit) buffering that cause this _using()_ to run cleanup before the consumption of that buffer. 
 
Is this correct?

Just thinking loud since doFinally used to keep the Response open until the full chain terminates and any impact by new change when deep-copy-disabled:true.

Reply:
Shouldn't the `Flux.using` result in effectively the same lifetime for the response? Where both cases would close the response shortly after a terminal event from the `response.bodyIntern` stream.

Reply:
I was thinking about the following use case (if it's ever valid when deep-copy-disabled:true). Like shown in the output the release of response (owning the buffer?) happens early in case of `using`, so I was wondering if this is valid execution in our Netty case with deep-copy-disabled:true, if so, could this lead to data loss / ccorruption. 

### Cleanup with doFinally operator

```java
@Test
public void cleanupViaDoFinally() throws InterruptedException {
    Flux<Integer> nioFlux = Flux.range(1, 10)
        .publishOn(Schedulers.boundedElastic()); // mimic emission of nio buffer in netty-io thread

    CountDownLatch latch = new CountDownLatch(1);
    nioFlux
        .doFinally(s -> {
            System.out.println("running 'doFinally' to close-response.");
        })
        .collectList() // mimic buffering of nio buffers
        .subscribe(value -> {
            try {
                TimeUnit.SECONDS.sleep(1);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("nio-buffers:" + value);
        }, e -> {
            System.out.println("DownStream: Got Error");
            latch.countDown();
        }, () -> {
            System.out.println("DownStream: Got Completion");
            latch.countDown();
        });

    latch.await();
}
```

Here the output is:

> nio-buffers:[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
DownStream: Got Completion
running 'doFinally' to close-response.

### Cleanup with using operator

```java
@Test
public void cleanupViaUsing() throws InterruptedException {
    Flux<Integer> sourceFlux = Flux.range(1, 10)
        .publishOn(Schedulers.boundedElastic()); // mimic emission of nio buffer in netty-io thread

    CountDownLatch latch = new CountDownLatch(1);
    Flux.using(() -> sourceFlux, s -> s, s -> { System.out.println("running 'using-cleanup' to close-response.");})
        .collectList()// mimic buffering of nio buffers
        .subscribe(value -> {
            try {
                TimeUnit.SECONDS.sleep(1);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("nio-buffers:" + value);
        }, e -> {
            System.out.println("DownStream: Got Error");
            latch.countDown();
        }, () -> {
            System.out.println("DownStream: Got Completion");
            latch.countDown();
        });

    latch.await();
}
```

Here the output is:

> running 'using-cleanup' to close-response.
nio-buffers:[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
DownStream: Got Completion

Reply:
I think this should be fine as `collectList` will collect all emissions from the `nioFlux` before the resource cleanup triggers. Putting that into perspective of the API in azure-core-http-netty, with buffering true or false either way the underlying network response will be consumed without ability for it to be replayed. When buffering the read data will be made durable by being brought into heap memory.

Reply:
Sounds great, in that case this is safe, thanks!.